Brokerages/WebSocketWrapper.cs

```diff
@@ -63,17 +55,110 @@ public void Initialize(string url)
         /// <param name="data"></param>
         public void Send(string data)
         {
-            _wrapped.Send(data);
+            try
+            {
+                if (!IsOpen)
+                {
+                    Log.Trace($"WebSocketWrapper.Send(): Connection not open (IsOpen:{IsOpen}, State:{_wrapped.State}): attempting connection", true);
+
+                    Connect();
+                }
+
+                var sendBuffer = new ArraySegment<byte>(Encoding.UTF8.GetBytes(data));
+
+                _wrapped.SendAsync(sendBuffer, WebSocketMessageType.Text, true, _cancellationTokenSource.Token).SynchronouslyAwaitTask();
+            }
+            catch (Exception exception)
+            {
+                OnError(new WebSocketError(exception.Message, exception));
+            }
         }
 
         /// <summary>
         /// Wraps Connect method
         /// </summary>
         public void Connect()
         {
-            if (!IsOpen)
+            if (IsOpen)
+            {
+                return;
+            }
+
+            try
+            {
+                _wrapped.ConnectAsync(new Uri(_url), _cancellationTokenSource.Token).SynchronouslyAwaitTask();
+
+                if (IsOpen)
+                {
+                    OnOpen();
+
+                    new TaskFactory().StartNew(
+                        async () =>
+                        {
+                            try
+                            {
+                                await StartReceiving();
+                            }
+                            catch (Exception exception)
+                            {
+                                OnError(new WebSocketError(exception.Message, exception));
+                            }
+
+                            if (_wrapped.State == WebSocketState.Aborted)
+                            {
+                                // need to recreate ClientWebSocket instance, cannot be reused
+
+                                Log.Trace($"WebSocketWrapper.Connect(): Connection aborted (IsOpen:{IsOpen}, State:{_wrapped.State}): creating new ClientWebSocket instance", true);
+
+                                await new TaskFactory().StartNew(ResetAndConnect, _cancellationTokenSource.Token);
```

Comment:
why did we choose to use `new TaskFactory()` instead of `TaskFactory.StartNew` or simply `Task.Run` ?

>https://devblogs.microsoft.com/pfxteam/task-run-vs-task-factory-startnew/

Even stranger though, `ResetAndConnect` is a synchronous method but we're wrapping it in a task _and_ awaiting it.... why?